src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,320 @@
+    /// A singleton that tracks the open IVsWindowFrames and can report which documents are visible or active in a given <see cref="Workspace"/>.
+    /// Can be accessed via the <see cref="IDocumentTrackingService"/> as a workspace service.
+    /// </summary>
+    [Export]
+    internal class VisualStudioActiveDocumentTracker : ForegroundThreadAffinitizedObject, IVsSelectionEvents, IDisposable
+    {
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+
+        /// <summary>
+        /// Collection of all asynchronous tasks that are started by this service. This should only be tasks that are implicitly
+        /// async as we fetch services from the <see cref="IAsyncServiceProvider"/>, and are used to ensure we wait for them during shutdown.
+        /// These should not be waited for in calls to <see cref="TryGetActiveDocument(Workspace)"/> or <see cref="GetVisibleDocuments(Workspace)"/>
+        /// because those are expected to not be jumping to the UI thread per traditional Roslyn semantics and may deadlock.
+        /// </summary>
+        private readonly JoinableTaskCollection _asyncTasks;
+
+        /// <summary>
+        /// The list of tracked frames. This can only be written by the UI thread, although can be read (with care) from any thread.
+        /// </summary>
+        private ImmutableList<FrameListener> _visibleFrames = ImmutableList<FrameListener>.Empty;
+
+        /// <summary>
+        /// The active IVsWindowFrame. This can only be written by the UI thread, although can be read (with care) from any thread.
+        /// </summary>
+        private IVsWindowFrame _activeFrame;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioActiveDocumentTracker(
+            IThreadingContext threadingContext,
+            [Import(typeof(SVsServiceProvider))] IAsyncServiceProvider asyncServiceProvider,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService)
+            : base(threadingContext, assertIsForeground: false)
+        {
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _asyncTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
+            _asyncTasks.Add(ThreadingContext.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var monitorSelectionService = (IVsMonitorSelection)await asyncServiceProvider.GetServiceAsync(typeof(SVsShellMonitorSelection));
+
+                if (ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_DocumentFrame, out var value)))
+                {
+                    if (value is IVsWindowFrame windowFrame)
+                    {
+                        TrackWindowFrame(windowFrame);
+                        _activeFrame = windowFrame;
```

Comment:
do we need to fire `this.DocumentsChanged?.Invoke(this, EventArgs.Empty);` here?

Reply:
Ah, right, changed to a single helper.